src/core/search/search_test.cc

```diff
@@ -0,0 +1,279 @@
+
+namespace dfly {
+namespace search {
+
+using namespace std;
+
+struct MockedDocument : public DocumentAccessor {
+ public:
+  using Map = absl::flat_hash_map<std::string, std::string>;
+
+  MockedDocument() = default;
+  MockedDocument(Map map) : fields_{map} {
+  }
+  MockedDocument(std::string test_field) : fields_{{"field", test_field}} {
+  }
+
+  bool Check(DocumentAccessor::FieldConsumer f, string_view active_field) const override {
+    if (!active_field.empty()) {
+      auto it = fields_.find(string{active_field});
+      return f(it != fields_.end() ? it->second : "");
+    } else {
+      for (const auto& [k, v] : fields_) {
+        if (f(v))
+          return true;
+      }
+      return false;
+    }
+  }
+
+  string DebugFormat() {
+    string out = "{";
+    for (const auto& [field, value] : fields_)
+      absl::StrAppend(&out, field, "=", value, ",");
+    if (out.size() > 1)
+      out.pop_back();
+    out += "}";
+    return out;
+  }
+
+  void Set(Map hset) {
+    fields_ = hset;
+  }
+
+ private:
+  Map fields_{};
+};
+
+// Just a filler. TODO: Remove
+struct Schema {
+  enum FieldType { TEXT, NUMERIC };
+
+  absl::flat_hash_map<std::string, FieldType> fields;
+};
+
+class SearchParserTest : public ::testing::Test {
+ protected:
+  SearchParserTest() {
+    PrepareSchema();
+  }
+
+  ~SearchParserTest() {
+    EXPECT_EQ(entries_.size(), 0u) << "Missing check";
+  }
+
+  void PrepareSchema(Schema schema = {{{"field", Schema::TEXT}}}) {
+    schema_ = schema;
+  }
+
+  void PrepareQuery(string_view query) {
+    query_ = query;
+  }
+
+  template <typename... Args> void ExpectAll(Args... args) {
+    (entries_.emplace_back(args, true), ...);
+  }
+
+  template <typename... Args> void ExpectNone(Args... args) {
+    (entries_.emplace_back(args, false), ...);
+  }
+
+  bool Check() {
+    SearchAlgorithm search_algo{};
+    search_algo.Init(query_);
+
+    for (auto& [doc, expect_match] : entries_) {
+      bool doc_matched = search_algo.Check(&doc);
+
+      if (doc_matched != expect_match) {
+        error_ = "doc: \"" + doc.DebugFormat() + "\"" + " was expected" +
+                 (expect_match ? "" : " not") + " to match" + " query: \"" + query_ + "\"";
+        entries_.clear();
+        return false;
+      }
+    }
+
+    entries_.clear();
+    return true;
+  }
+
+  string_view GetError() const {
+    return error_;
+  }
+
+ private:
+  using DocEntry = pair<MockedDocument, bool /*should_match*/>;
+
+  Schema schema_;
+  vector<DocEntry> entries_;
+  string query_, error_;
+};
+
+TEST_F(SearchParserTest, MatchTerm) {
+  PrepareQuery("foo");
+
+  // Check basic cases
+  ExpectAll("foo", "foo bar", "more foo bar");
+  ExpectNone("wrong", "nomatch");
+
+  // Check part of sentence + case.
+  ExpectAll("Foo is cool.", "Where is foo?", "One. FOO!. More", "Foo is foo.");
+
+  // Check part of word is not matched
+  ExpectNone("foocool", "veryfoos", "ufoo", "morefoomore", "thefoo");
+
+  EXPECT_TRUE(Check()) << GetError();
```

Comment:
I now assert only at the very end because I want all documents to be contained in the index, so it performs more complex operations. 

Yes, this is not idiomatic "one assert per new test case". The alternative is building the index ahead and then running different queries on it and asserting on those (like in the search family test)

This would require require rewriting all the tests to reverse the relation. Not sure it's worth it now